lonestar/analytics/cpu/skip-gram/Embedding.cpp

```diff
@@ -0,0 +1,221 @@
+#include <algorithm>
+#include <array>
+#include <fstream>
+#include <iostream>
+#include <map>
+#include <set>
+#include <sstream>
+#include <string>
+#include <vector>
+
+#include "Huffman/HuffmanCoding.h"
+#include "Lonestar/BoilerPlate.h"
+#include "NeuralNetwork/SkipGramModelTrainer.h"
+
+namespace cll = llvm::cl;
+
+static const char* name = "Embeddings";
+static const char* desc = "Generate embeddings";
+static const char* url = "embeddings";
+
+static cll::opt<std::string> inputFile(
+    cll::Positional, cll::desc("<input file>"), cll::Required);
+
+static cll::opt<std::string> outputFile(
+    cll::Positional, cll::desc("<output file>"), cll::Required);
+
+static cll::opt<uint32_t> embeddingSize(
+    "embeddingSize",
+    cll::desc("Size of the embedding vector (default value 100)"),
+    cll::init(100));
+
+static cll::opt<double> alpha(
+    "alpha", cll::desc("alpha (default value 0.025)"), cll::init(0.025f));
+
+static cll::opt<uint32_t> window(
+    "window", cll::desc("window size (default value 5)"), cll::init(5));
+
+static cll::opt<double> downSampleRate(
+    "downSampleRate", cll::desc("down-sampling rate (default value 0.001)"),
+    cll::init(0.001f));
+
+static cll::opt<bool> hierarchicalSoftmax(
+    "hierarchicalSoftmax",
+    cll::desc("Enable/disable hierarchical softmax (default value false)"),
+    cll::init(false));
+
+static cll::opt<uint32_t> numNegSamples(
+    "numNegSamples", cll::desc("Number of negative samples (default value 5)"),
+    cll::init(5));
+
+static cll::opt<uint32_t> numIterations(
+    "numIterations",
+    cll::desc("Number of Training Iterations (default value 5)"), cll::init(5));
+
+static cll::opt<uint32_t> minCount(
+    "minCount", cll::desc("Min-count (default 5)"), cll::init(5));
+
+void
+ReadRandomWalks(
+    std::ifstream& input_file,
+    std::vector<std::vector<uint32_t>>* random_walks) {
+  std::string line;
+
+  while (std::getline(input_file, line)) {
+    std::vector<uint32_t> walk;
+    std::stringstream ss(line);
+
+    uint32_t val;
+
+    while (ss >> val) {
+      walk.push_back(val);
+    }
+
+    random_walks->push_back(std::move(walk));
+  }
+}
+
+void
+BuildVocab(
```

Comment:
Parallelize this function. Use katana::gstl::Map<uint32_t, uint32_t>. See CC for example: https://github.com/KatanaGraph/katana/blob/master/libgalois/src/analytics/connected_components/connected_components.cpp